### PR TITLE
feat(server): implement json.arrindex command (#104)

### PR DESCRIPTION
**Pull request by iko1**
Signed-off-by: iko1 <me@remotecpp.dev>

### CODE REVIEW

**Comment by romange** [comment]
please rebase

**Comment by iko1** [comment]
> 

Done.

**Comment by iko1** [comment]
> please rebase

Done.

File: src/server/json_family.cc [hunk 1]
```diff
@@ -692,8 +702,135 @@ OpResult<vector<OptSizeT>> OpArrTrim(const OpArgs& op_args, string_view key, str
...
+    }
+
+    if (val.empty()) {
+      vec.emplace_back(-1);
+      return;
+    }
+
+    if (start_index < 0) {
+      int idx = val.size() + start_index;
+      if (idx < 0) {
```
Comment (romange) [inline src/server/json_family.cc:729]: what's the point of doing `idx %= val.size()` for negative indices?  it's not a common op, so some comments would be useful.

File: src/server/json_family.cc [hunk 2]
```diff
@@ -692,8 +702,135 @@ OpResult<vector<OptSizeT>> OpArrTrim(const OpArgs& op_args, string_view key, str
...
+        idx %= val.size();
+      }
+
+      search_start_index = abs(idx);
+    } else {
+      search_start_index = start_index;
+    }
+
+    if (end_index <= 0) {
+      if (end_index == 0 || end_index == -1) {
```
Comment (romange) [inline src/server/json_family.cc:739]: maybe I can help with combining those if's if you write here what is the expected mapping for negative values.

File: src/server/json_family.cc [hunk 3]
```diff
@@ -692,8 +702,135 @@ OpResult<vector<OptSizeT>> OpArrTrim(const OpArgs& op_args, string_view key, str
   return vec;
 }
 
+OpResult<vector<OptLong>> OpArrIndex(const OpArgs& op_args, string_view key,
+                                     JsonExpression expression, const json& search_val,
```
Comment (romange) [inline src/server/json_family.cc:706]: can we search for any json object? even for an object?

Comment (iko1) [reply src/server/json_family.cc:706]: No. according to Elasticache documentation, only JSON scalar is searchable.
"JSON scalar refers to values that are not objects or arrays. i.e., String, number, boolean and null are scalar values."

File: src/server/json_family.cc (hunk 2 above)
Comment (iko1) [reply src/server/json_family.cc:739]: Negative index values are treated by rounding the index to the array's start and end.

File: src/server/json_family.cc (hunk 3 above)
Comment (romange) [reply src/server/json_family.cc:706]: please add a comment quoting this.

File: src/server/json_family.cc (hunk 2 above)
Comment (romange) [reply src/server/json_family.cc:739]: Please add a comment at line 728 with example on how say -2 is mapped when val.size() is 9.
and how -11 should map.

I think it's possible to improve the code. at line 728:
```cpp
if (start_index < 0) {
  start_index %= val.size();  // start_index stays negative but greater than -val.size().
  start_index += val.size();  // 0 < start_index < val.size()
}
```
no need search_start_index. Similarly with end_index.

File: src/server/json_family.cc [hunk 4]
```diff
@@ -771,8 +772,126 @@ OpResult<vector<OptSizeT>> OpArrInsert(const OpArgs& op_args, string_view key, s
...
+
+  if (search_value->is_object() || search_value->is_array()) {
+    (*cntx)->SendError(kWrongTypeErr);
+    return;
+  }
+
+  int start_index = 0;
+  if (args.size() >= 5) {
+    if (!absl::SimpleAtoi(ArgS(args, 4), &start_index)) {
+      VLOG(1) << "Failed to convert the start index to numeric" << ArgS(args, 4);
```
Comment (romange) [inline src/server/json_family.cc:869]: why you do not return the error if parsing failed? here and with `end_index` ?

**Review by romange — COMMENTED** [review]
mostly lgtm

File: src/server/json_family.cc (hunk 4 above)
Comment (iko1) [reply src/server/json_family.cc:869]: actually, I thought about that before, but then I check what happened in Redis Implementation of JSON, and I found they ignore completely those index values and proceed with the search operation if there is a parse error.

Comment (romange) [reply src/server/json_family.cc:869]: I do not want to mirror this behavior. You can add a comment saying that it differs from redis but I prefer to reject incorrect arguments similarly to how it's done with other commands.

Comment (iko1) [reply src/server/json_family.cc:869]: Fixed.

**Review by romange — APPROVED** [review]

**Pull request merged**